Add DebugSSAUpdater class to track debug value liveness

SLTozer opened the pull request:
This patch adds a class that uses SSA construction, with debug values as definitions, to determine whether and which debug values for a particular variable are live at each point in an IR function. This will be used by the IR reader of llvm-debuginfo-analyzer to compute variable ranges and coverage, although it may be applicable to other debug info IR analyses.

SLTozer committed:
Add DebugSSAUpdater class to track debug value liveness

This patch adds a class that uses SSA construction, with debug values as
definitions, to determine whether and which debug values for a particular
variable are live at each point in an IR function. This will be used by the
IR reader of llvm-debuginfo-analyzer to compute variable ranges and
coverage, although it may be applicable to other debug info IR analyses.

jmorse commented on llvm/lib/Transforms/Utils/DebugSSAUpdater.cpp:
```diff
@@ -0,0 +1,415 @@
+    LLVM_DEBUG(dbgs() << "Create set of ranges with " << BlockDbgRanges.size()
+                      << " entries!\n");
+    if (!BlockDbgRanges.empty())
+      OrigVariableValueRangeTable[DVA].append(BlockDbgRanges);
```
I feel some kind of move-semantics appending / emplacing / whatever here is going to be worth installing for speed purposes, or just on principle.

jmorse commented on llvm/lib/Transforms/Utils/DebugSSAUpdater.cpp:
```diff
@@ -0,0 +1,415 @@
+  }
+}
+
+uint64_t DbgValueRangeTable::addVariableName(Value *V, uint64_t Size) {
```
Not called by anything -- is this unused? (It's also unclear what it's doing).

jmorse reviewed (commented):
This is looking good with various nits inline. I didn't look at the unit test as I ran out of time. Rhyme.

Various bits here are duplicated from SSAUpdater, which I believe is fine as trying to abstract it further would probably be a false economy. Some kind of comment is in order to signal to readers where it's duplicated from, IMO, not sure where or how to word it though.

SLTozer committed:
Address review comments, split out value-name mapping to separate class, remove line tracking

SLTozer committed:
clang-format

SLTozer commented:
Rebased the patch onto main and addressed review comments; also split out functionality from the DbgValueRangeTable that isn't related to variable location ranges. The line and name tracking features are used by the debug info analyzer's IR reader, but they don't need to be part of this class - however, I have moved the name tracking feature to a separate class, `SSAValueNameMap`, and added comments to describe its purpose. Since it's not strictly related to the liveness analysis it could be moved out of this file and review, but it's a utility class that is useful to the only consumer of this patch, so I think we could reasonably keep it around - happy to adjust the patch if not.

jmorse reviewed (approved):
LGTM

SLTozer committed:
Use right type for test comp
